components/raftstore/src/coprocessor/region_info_accessor.rs

```diff
@@ -709,26 +741,80 @@ impl RegionInfoProvider for RegionInfoAccessor {
                 })
             })
     }
+
+    fn find_region_by_key(&self, key: &[u8]) -> Result<Option<Region>> {
+        let (tx, rx) = mpsc::channel();
+        let msg = RegionInfoQuery::GetRegionByKey {
+            key: key.to_vec(),
+            callback: Box::new(move |region| {
+                if let Err(e) = tx.send(region) {
+                    warn!("failed to send get_regions_in_range result: {:?}", e);
+                }
+            }),
+        };
+        self.scheduler
+            .schedule(msg)
+            .map_err(|e| box_err!("failed to send request to region collector: {:?}", e))
+            .and_then(|_| {
+                rx.recv().map_err(|e| {
+                    box_err!(
+                        "failed to receive find_region_by_key result from region collector: {:?}",
+                        e
+                    )
+                })
+            })
+    }
 }
 
 // Use in tests only.
-pub struct MockRegionInfoProvider(Mutex<Vec<Region>>);
+// Note: The `StateRole` in RegionInfo here should not be used
+pub struct MockRegionInfoProvider(Mutex<Vec<RegionInfo>>);
 
 impl MockRegionInfoProvider {
     pub fn new(regions: Vec<Region>) -> Self {
-        MockRegionInfoProvider(Mutex::new(regions))
+        MockRegionInfoProvider(Mutex::new(
+            regions
+                .into_iter()
+                .map(|region| RegionInfo::new(region, StateRole::Leader))
+                .collect_vec(),
+        ))
     }
 }
 
 impl Clone for MockRegionInfoProvider {
     fn clone(&self) -> Self {
-        MockRegionInfoProvider::new(self.0.lock().unwrap().clone())
+        MockRegionInfoProvider::new(
+            self.0
+                .lock()
+                .unwrap()
+                .clone()
+                .into_iter()
+                .map(|region_info| region_info.region)
+                .collect_vec(),
+        )
     }
 }
 
 impl RegionInfoProvider for MockRegionInfoProvider {
     fn get_regions_in_range(&self, _start_key: &[u8], _end_key: &[u8]) -> Result<Vec<Region>> {
-        Ok(self.0.lock().unwrap().clone())
+        Ok(self
+            .0
+            .lock()
+            .unwrap()
+            .clone()
+            .into_iter()
+            .map(|region_info| region_info.region)
+            .collect_vec())
+    }
+
+    fn seek_region(&self, from: &[u8], callback: SeekRegionCallback) -> Result<()> {
+        let region_infos = self.0.lock().unwrap();
+        let mut iter = region_infos.iter().filter(|&region_info| {
+            RangeKey::from_start_key(region_info.region.get_start_key().to_vec())
+                <= RangeKey::from_start_key(from.to_vec())
```

Comment:
It seems that you returned the region that contains `from` and all regions *before* `from`, however we expects the region that contains `from` and all regions *after* `from`. I think what you need is:

```suggestion
            RangeKey::from_end_key(region_info.region.get_end_key().to_vec())
                > RangeKey::from_start_key(from.to_vec())
```

Reply:
Actually I think it might be a good idea to add some simple tests for it, though it's a mock.

Reply:
Good suggestion.